bindings/python/tests/pire_test.py

```diff
@@ -0,0 +1,401 @@
+        pytest.raises(Exception, scanner_class.Load, saved[1:])
+
+    def test_scanner_finds_prefixes_and_suffixes(self, scanner_class):
+        fsm = pire.Lexer("-->").Parse()
+        any_occurence = scanner_class(~pire.Fsm.MakeFalse() + fsm)
+        first_occurence = scanner_class(~fsm.Surrounded() + fsm)
+        reverse_occurence = scanner_class(fsm.Reverse())
+
+        text = "1234567890 --> middle --> end"
+        assert 14 == first_occurence.LongestPrefix(text)
+        assert 11 == reverse_occurence.LongestSuffix(text[:14])
+
+        assert 25 == any_occurence.LongestPrefix(text)
+        assert 22 == reverse_occurence.LongestSuffix(text[:25])
+
+        assert 14 == first_occurence.ShortestPrefix(text)
+        assert 11 == reverse_occurence.ShortestSuffix(text[:14])
+
+        assert 14 == any_occurence.ShortestPrefix(text)
+        assert 11 == reverse_occurence.ShortestSuffix(text[:14])
+
+    def test_scanner_does_not_find_nonexistent_prefixes_and_suffixes(self, parse_scanner):
+        scanner = parse_scanner("text")
+        assert None is scanner.LongestPrefix("nonexistent")
+        assert None is scanner.ShortestPrefix("nonexistent")
+        assert None is scanner.LongestSuffix("nonexistent")
+        assert None is scanner.ShortestSuffix("nonexistent")
+
+    def test_glued_scanners_have_runnable_state(self, scanner_class, parse_scanner):
+        scanners_without_glue = [
+            pire.SimpleScanner,
+            pire.SlowScanner,
+            pire.CapturingScanner,
+        ]
+        if scanner_class in scanners_without_glue:
+            return
+
+        glued = parse_scanner("ab").GluedWith(parse_scanner("abcd$"))
+
+        assert 2 == glued.RegexpsCount()
+
+        state = glued.InitState()
+        check_state(state, final=False, dead=False)
+        check_state(state.Run("ab"), final=True, accepted_regexps=(0,))
+        check_state(state.Run("cd"), final=False, dead=False)
+        check_state(state.End(), final=True, accepted_regexps=(1,))
+        check_state(state.Run("-"), final=False, dead=True)
+
+        doubled = glued.GluedWith(glued)
+
+        state = doubled.InitState()
+        check_state(state.Run("ab"), final=True, accepted_regexps=(0, 2))
+        check_state(state.Run("cd").End(), final=True, accepted_regexps=(1, 3))
+
+    def test_state_remembers_its_scanner(self, scanner_class):
+        scanner = scanner_class()
+        state = scanner.InitState()
+        assert state.scanner == scanner
+
+
+class TestEasy(object):
+    def test_options_have_only_default_constuctor(self):
+        pire.Options()
+        pytest.raises(Exception, pire.Options, 1)
+
+    def test_regexp_matches(self):
+        re = pire.Regexp("(foo|bar)+", pire.I)
+        assert "prefix fOoBaR suffix" in re
+        assert "bla bla bla" not in re
+        assert re.Matches("barfoo")
+
+    def test_regexp_honors_utf8(self):
+        re = pire.Regexp("^.$", pire.I | pire.UTF8)
+        assert "\x41" in re
```

Comment:
What is really important is that "\x41" (or "A") is a valid UTF-8. From "A" it is clearer.